### PR TITLE
Adding Data Node Test Implemented #407

### PR DESCRIPTION
**Pull request by tsuchiyama-araya**
Adding Data Node Test Implemented #407

Contents:
- mocked store to pass into the component. Then tested if it shows properly
- Tested when the Data Nodes is Clicked then Image clicked it will verify that the action was dispatched with the expected payload

AddOns -> Fixed Typo for Function Name.

All Tests Run After Implementing New Test:
Screenshot 2024-09-02 at 14 20 25

### CODE REVIEW

**Comment by tsuchiyama-araya** [comment]
How to do the test

1.) git checkout feature/#407_add_data_node_test
2.) go to frontend directory
3.) run yarn install
4.) run yarn run test

It should run the tests. All of it should be passing.

**Comment by milesAraya** [comment]
Seems to work correctly
`yarn install`
`yarn run test`
Output:

> (node:74770) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
> (Use `node --trace-deprecation ...` to show where the warning was created)
>  PASS  src/store/slice/AlgorithmList/AlgorithmList.test.ts
>  PASS  src/store/test/visualize/Visualize.test.ts
> (node:74769) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
> (Use `node --trace-deprecation ...` to show where the warning was created)
> (node:74772) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
> (Use `node --trace-deprecation ...` to show where the warning was created)
>  PASS  src/store/test/RunSelectors.test.ts (5.536 s)
>  PASS  src/store/test/record/RecordReproduce.test.ts (5.537 s)
>  PASS  src/store/slice/Experiments/Experiments.test.ts
>  PASS  src/store/test/workflow/AlgorithmNode/AlgorithmNode.test.ts
> One of your dependencies, babel-preset-react-app, is importing the
> "@babel/plugin-proposal-private-property-in-object" package without
> declaring it in its dependencies. This is currently working because
> "@babel/plugin-proposal-private-property-in-object" is already in your
> node_modules folder for unrelated reasons, but it may break at any time.
> 
> babel-preset-react-app is part of the create-react-app project, which
> is not maintianed anymore. It is thus unlikely that this bug will
> ever be fixed. Add "@babel/plugin-proposal-private-property-in-object" to
> your devDependencies to work around this error. This will make this message
> go away.
>   
>  PASS  src/store/test/workflow/InputNode/InputNode.test.ts
> One of your dependencies, babel-preset-react-app, is importing the
> "@babel/plugin-proposal-private-property-in-object" package without
> declaring it in its dependencies. This is currently working because
> "@babel/plugin-proposal-private-property-in-object" is already in your
> node_modules folder for unrelated reasons, but it may break at any time.
> 
> babel-preset-react-app is part of the create-react-app project, which
> is not maintianed anymore. It is thus unlikely that this bug will
> ever be fixed. Add "@babel/plugin-proposal-private-property-in-object" to
> your devDependencies to work around this error. This will make this message
> go away.
>   
>  PASS  src/store/slice/FilesTree/FilesTree.test.ts
> One of your dependencies, babel-preset-react-app, is importing the
> "@babel/plugin-proposal-private-property-in-object" package without
> declaring it in its dependencies. This is currently working because
> "@babel/plugin-proposal-private-property-in-object" is already in your
> node_modules folder for unrelated reasons, but it may break at any time.
> 
> babel-preset-react-app is part of the create-react-app project, which
> is not maintianed anymore. It is thus unlikely that this bug will
> ever be fixed. Add "@babel/plugin-proposal-private-property-in-object" to
> your devDependencies to work around this error. This will make this message
> go away.
>   
> (node:74771) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
> (Use `node --trace-deprecation ...` to show where the warning was created)
> One of your dependencies, babel-preset-react-app, is importing the
> "@babel/plugin-proposal-private-property-in-object" package without
> declaring it in its dependencies. This is currently working because
> "@babel/plugin-proposal-private-property-in-object" is already in your
> node_modules folder for unrelated reasons, but it may break at any time.
> 
> babel-preset-react-app is part of the create-react-app project, which
> is not maintianed anymore. It is thus unlikely that this bug will
> ever be fixed. Add "@babel/plugin-proposal-private-property-in-object" to
> your devDependencies to work around this error. This will make this message
> go away.
>   
>  PASS  src/components/Workspace/FlowChart/TreeView.test.tsx (9.943 s)
> 
> Test Suites: 9 passed, 9 total
> Tests:       27 passed, 27 total
> Snapshots:   0 total
> Time:        11.883 s
> Ran all test suites.

**Comment by milesAraya** [comment]
Maybe this is a good opportunity to update yarn to include
`@babel/plugin-proposal-private-property-in-object`
and use an alternative to `punycode`

**Review by milesAraya — APPROVED** [review]
Maybe this is a good opportunity to update yarn to include
`@babel/plugin-proposal-private-property-in-object`
and use an alternative to `punycode`

**Comment by tsuchiyama-araya** [comment]
@milesAraya Thank you for checking. I will try to update yarn.

**Comment by tsuchiyama-araya** [comment]
@milesAraya 
お疲れ様です。
@babel/plugin-proposal-private-property-in-object　に関しては対応しましたが、punycodeについてはなかなか解決はできないため、一旦放置しています。
現状は優先するものではないかと思います。

File: frontend/src/components/Workspace/FlowChart/TreeView.test.tsx [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+    render(
+      <Provider store={store}>
+        <AlgorithmTreeView />
+      </Provider>,
+    )
+
+    expect(store.dispatch).toHaveBeenCalledWith(getAlgoList())
+  })
+
+  it("dispatches the correct action when the add button is clicked", async () => {
```
[the remark this answers is not in the document]
Comment (tsuchiyama-araya) [reply frontend/src/components/Workspace/FlowChart/TreeView.test.tsx:409]: はい、おっしゃる通りImage Nodeを１つ追加するようになっています。
もう少し、具体的に書いてみます。

File: frontend/package.json [hunk 2]
```diff
@@ -98,6 +98,7 @@
     "lint-staged": "^15.0.2",
     "prettier": "^3.0.3",
     "process": "^0.11.10",
+    "punycode": "^2.3.1",
```
[the remark this answers is not in the document]
Comment (tsuchiyama-araya) [reply frontend/package.json:101]: こちらなんですが、一応追加しましたが、
それでもまだ、PunyCodeのWarningメッセージは表示されているので、完全解決ではないかと思います。
外した方がいいですかね？

[the remark this answers is not in the document]
Comment (itutu-tienday) [reply frontend/package.json:101]: こちらの環境では、warningは特に出ていないようです。 （以下screenshot）
そちらはどの様な状態でしょうか？

image

File: frontend/src/components/Workspace/FlowChart/TreeView.test.tsx [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+
+const mockStore = configureStore([thunk])
+
+describe("AlgorithmTreeView", () => {
+  let store: ReturnType<typeof mockStore>
+
+  beforeEach(() => {
+    store = mockStore({
+      isLatest: true,
+      algorithmList: {
```
[the remark this answers is not in the document]
Comment (tsuchiyama-araya) [reply frontend/src/components/Workspace/FlowChart/TreeView.test.tsx:32]: テストデータは、別ファイルにしてみます。
このデータを使い回しをする可能性がありそうなので、分離した方が良さそうです。

File: frontend/package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (tsuchiyama-araya) [reply frontend/package.json:101]: 私の場合はこのように表示されていますので、
Warningは出ています。環境の問題でしょうかね。。

一応、yarn run test ＆ npm run test で試しても、同じ結果でした。

Screenshot 2024-09-10 at 16 44 51

[the remark this answers is not in the document]
Comment (itutu-tienday) [reply frontend/package.json:101]: punycode を追加しても、warningに変化がないのであれば、package.json への追加は不要で良さそうですね。

[the remark this answers is not in the document]
Comment (tsuchiyama-araya) [reply frontend/package.json:101]: 特に副作用はないかと思います。

[the remark this answers is not in the document]
Comment (tsuchiyama-araya) [reply frontend/package.json:102]: @itutu-tienday こちらは解決にして大丈夫でしょうか？

File: frontend/src/components/Workspace/FlowChart/TreeView.test.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (itutu-tienday) [reply frontend/src/components/Workspace/FlowChart/TreeView.test.tsx]: @tsuchiyama-araya 
修正を適用されましたら、再レビュー依頼をお知らせください。
※Conversations の Resolve は、レビュー指摘者が再チェック後に、Resolveするフローの想定です。

File: frontend/src/components/Workspace/FlowChart/testdata/mockStoreData.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            args: [{ name: "image", type: "ImageData", isNone: false }],
+            returns: [{ name: "ops", type: "Suite2pData" }],
+          },
+          suite2p_registration: {
+            type: "child",
+            functionPath: "suite2p/suite2p_registration",
+            args: [{ name: "ops", type: "Suite2pData", isNone: false }],
+            returns: [{ name: "ops", type: "Suite2pData" }],
+          },
+          // 残りのデータも同様に記述
```
Comment (itutu-tienday) [inline frontend/src/components/Workspace/FlowChart/testdata/mockStoreData.ts:40]: このコメントは、どのような内容でしょうか？

File: frontend/src/components/Workspace/FlowChart/testdata/mockStoreData.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
+export const mockStoreData = {
```
Comment (itutu-tienday) [inline frontend/src/components/Workspace/FlowChart/testdata/mockStoreData.ts:1]: これは質問ですが、テストデータは json 形式ではなく、 ts コード形式が適当そうでしょうか？

- json ファイルの import
  - https://www.typescriptlang.org/tsconfig/#resolveJsonModule

File: frontend/src/components/Workspace/FlowChart/TreeView.test.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (tsuchiyama-araya) [reply frontend/src/components/Workspace/FlowChart/TreeView.test.tsx]: @itutu-tienday 
ご指摘ありがとうございます。
テストデータなので、Json形式が正しいです。
失礼しました。

File: frontend/src/components/Workspace/FlowChart/testdata/mockStoreData.ts (hunk 4 above)
Comment (tsuchiyama-araya) [reply frontend/src/components/Workspace/FlowChart/testdata/mockStoreData.ts]: こちらのコメントを削除しました。
ご指摘ありがとうございます。

File: frontend/src/components/Workspace/FlowChart/testdata/mockStoreData.ts (hunk 5 above)
Comment (tsuchiyama-araya) [reply frontend/src/components/Workspace/FlowChart/testdata/mockStoreData.ts]: 指摘ありがとうございます。
Jsonファイルに修正しました。
お手数ですが、ご確認お願いいたします。

File: frontend/src/components/Workspace/FlowChart/TreeView.test.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (itutu-tienday) [reply frontend/src/components/Workspace/FlowChart/TreeView.test.tsx]: メリットがあれば、テストデータは .ts 形式でもOKなのかと思います。
いずれの形式が適当か、検討いただければと思います。
※現状は .json とする方針ということで、了解しました。

File: frontend/src/components/Workspace/FlowChart/TreeView.test.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (tsuchiyama-araya) [reply frontend/src/components/Workspace/FlowChart/TreeView.test.tsx]: @itutu-tienday 
こちら、改修しました。何がクリックされているかを記載しました。
お手数ですが、ご確認お願い致します。

**Review by itutu-tienday — APPROVED** [review]

**Pull request merged**